src/common/signal_handling.cpp

```diff
@@ -7,15 +7,27 @@
 
 namespace marian{
 volatile std::sig_atomic_t sigflags_{0};
+volatile std::sig_atomic_t gracefulExitRequested_{0};
 
 bool getSignalFlag(const int sig) {
   // sig_atomic_t has 32 bits. We don't accommodate signals beyond that.
-  ABORT_IF(sig >= 32, "Signal {} out of range (must be < 32).", sig);
+  ABORT_IF(sig >= 32, "Signal out of range (must be < 32, is {}).", sig);
   return sigflags_ & (1<<sig);
 }
 
+void requestGracefulExit(int sig) {
+  setSignalFlag(sig);         // keep track of triggering signal
+  gracefulExitRequested_ = 1; // set flag to exit gracefully
+  LOG(debug, "Graceful exit requested via signal {}.", sig);
```

Comment:
let's go for `info` instead of `debug`. I would say actually forcing an exit should always be logged.

Reply:
Technically, the graceful shutdown is being logged anyway (see Scheduler::finished()). I have no strong opinions about this, though. Also technically, the only thing that should happen in a signal handler is setting flags. as it's entirely unpredictable when a signal is handled. I don't know if the logger involves a lock, (and whether that lock has a timeout), but theoretically: if the signal handler is triggered while the logger is logging something else in the same thread, we might end up with a deadlock. Happy to take the logging out here (or comment it out should we ever need to debug again). Or we might guard this to be included only in the Debug build type but not in Release.

Reply:
I think the logging is thread safe.

Reply:
I'm with @ugermann here, do as little as possible in a signal handler.  You're limited to these functions https://man7.org/linux/man-pages/man7/signal-safety.7.html and would need to show that the logger only uses them as well as being reentrant in the same thread, not just thread safe.

Reply:
OK, if we just remove it then there is no question about the logger to use :)

Reply:
@emjotde: signal handling INTERRUPTS one thread (kernel decides which one among those that don't explicitly block the signal) to execute the signal handling in that thread, so even threat safety might not prevent a deadlock. The timing would have to be very unfortunate, but it is conceivable. Better not to mess with it. Logging removed from signal handler.